docs/design/design.md

```diff
@@ -0,0 +1,131 @@
+- **Peer**: In a P2P network, peers are both suppliers and consumers of resources. So before Dfget downloads a task from Supernode, Dfget will launch a webserver that provides downloaded files for other peers to download in the P2P network and send a `Peer/Register` request to Supernode to join the P2P network. Only after that can Dfget download files from P2P network.
+- **Piece**: A piece is a part of the file that is going to download, which can be interpreted as a block of files. In dragonfly, the downloading file is not transmitted in a whole file, but in pieces.
+
+## API Design
+
+For a program, a well-designed API is very important, we design the APIs with principles in `RESTFUL` as follows:
+
+- Keep each API's responsibilities simple.
+- Provide APIs that cover the full process life cycle. Now we've just exposed the APIs required by the core process, and we'll add the appropriate APIs as needed later.
+- Use resources as the heart of the API, but they do not have to correspond to an actual data object one-by-one. `E.g.` Tasks  are resources used by a client to operate, DfgetTasks are for internal implementation which should not be exposed to client.
+- URLs should include nouns, not verbs. In additional, use plural nouns instead of singular nouns. `E.g.` `/peers/{id}: GET`.
+- Use HTTP response status codes to represent the outcome of operations on resources. Please refer to [linker](https://github.com/dragonflyoss/Dragonfly/blob/master/apis/swagger.yml#L793).
+- Avoid URL's path level is too deep, we try not to have URLs deeper than resource/identifier/resource, the deepest URL in our APIs is /tasks/{id}/pieces/{pieceRange} because of special circumstances.
+- Well-documented,  we organize the API with swagger and keep the API synchronized with the documentation through swagger generation tools.
+
+### Reference Docs
+
+For API definition, please refer to [swagger.yml](https://github.com/dragonflyoss/Dragonfly/blob/master/apis/swagger.yml)
+
+For API document, please refer to [apis.md](https://github.com/dragonflyoss/Dragonfly/blob/master/docs/apis.md)
+
+## Architecture
+
+In DragonFly, we have two roles:
+
+- Supernode, a SuperNode is a long-time process that plays the following roles:
+  - A tracker and scheduler in the P2P network that chooses the appropriate downloading net-path for each peer.
+  - A CDN server that caches downloaded data from the source to avoid downloading same files repeatedly.
+- Dfclient, Dfclient contains two components:
+  - Dfget, Dfget is the Dragonfly client used for downloading files. It's similar to wget. Meanwhile, it also plays the role of a peer with uploader, which can transfer data between each other in the P2P network.
+  - Dfdaemon, Dfdaemon is only used for pulling images. It establishes a proxy between dockerd/pouchd and registry.Dfdaemon filters out layer fetching requests from all requests sent by dockerd/pouchd when pulling images, then uses Dfget to download these layers.
```

Comment:
`It establishes a proxy between dockerd/pouchd and registry.Dfdaemon` should be written as  `It establishes a proxy between dockerd/pouchd and Dfdaemon.registry`